File: ref/csharp/NelfeScoring.Vectors/Jcs.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace NelfeScoring;

/// <summary>
/// Canonicalisation JSON - RFC 8785 (JCS), sous-ensemble suffisant pour le contrat :
/// tous les nombres du passeport sont des ENTIERS (les valeurs sensibles à la
/// précision - score, temps - sont des chaînes). Les flottants sont donc refusés.
///
/// Règles : clés d'objet triées par unités de code UTF-16 (StringComparer.Ordinal),
/// aucun blanc, chaînes échappées à la ECMAScript JSON.stringify (échappements courts
/// \b \t \n \f \r, autres contrôles en \u00xx minuscule), non-ASCII laissé littéral.
/// </summary>
public static class Jcs
{
    public static byte[] CanonicalBytes(JsonNode? node)
        => Encoding.UTF8.GetBytes(Canonical(node));

    public static string Canonical(JsonNode? node)
    {
        // On repasse par le texte + JsonDocument pour un accès de type uniforme,
        // que le nœud vienne d'un parse ou d'une construction programmatique.
        using var doc = JsonDocument.Parse(node?.ToJsonString() ?? "null");
        var sb = new StringBuilder();
        Write(doc.RootElement, sb);
        return sb.ToString();
    }

    public static string Canonical(JsonElement el)
    {
        var sb = new StringBuilder();
        Write(el, sb);
        return sb.ToString();
    }

    private static void Write(JsonElement el, StringBuilder sb)
    {
        switch (el.ValueKind)
        {
            case JsonValueKind.Object:
                sb.Append('{');
                var first = true;
                foreach (var p in el.EnumerateObject().OrderBy(p => p.Name, StringComparer.Ordinal))
                {
                    if (!first) sb.Append(',');
                    first = false;
                    WriteString(p.Name, sb);
                    sb.Append(':');
                    Write(p.Value, sb);
                }
                sb.Append('}');
                break;

            case JsonValueKind.Array:
                sb.Append('[');
                var f2 = true;
                foreach (var item in el.EnumerateArray())
                {
                    if (!f2) sb.Append(',');
                    f2 = false;
                    Write(item, sb);
                }
                sb.Append(']');
                break;

            case JsonValueKind.String:
                WriteString(el.GetString()!, sb);
                break;

            case JsonValueKind.Number:
                if (!el.TryGetInt64(out var l))
                    throw new NotSupportedException(
                        $"Nombre non entier dans le passeport canonique : {el.GetRawText()}");
                sb.Append(l.ToString(CultureInfo.InvariantCulture));
                break;

            case JsonValueKind.True: sb.Append("true"); break;
            case JsonValueKind.False: sb.Append("false"); break;
            case JsonValueKind.Null: sb.Append("null"); break;
            default: throw new NotSupportedException($"ValueKind {el.ValueKind}");
        }
    }

    private static void WriteString(string s, StringBuilder sb)
    {
        sb.Append('"');
        foreach (var ch in s)
        {
            switch (ch)
            {
                case '"': sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                default:
                    if (ch < 0x20)
                        sb.Append("\\u").Append(((int)ch).ToString("x4", CultureInfo.InvariantCulture));
                    else
                        sb.Append(ch); // non-ASCII laissé littéral (UTF-8 à l'encodage)
                    break;
            }
        }
        sb.Append('"');
    }
}

File: ref/php/src/Jcs.php
<?php
declare(strict_types=1);

namespace NelfeScoring;

/**
 * Canonicalisation JSON - RFC 8785 (JCS), sous-ensemble du contrat (nombres ENTIERS
 * uniquement ; les valeurs sensibles sont des chaînes). DOIT produire des octets
 * IDENTIQUES à l'implémentation C# - sinon la signature ne vérifie pas.
 *
 * Décoder le passeport avec json_decode($json) (objets en stdClass, tableaux en
 * array) pour préserver la distinction objet/tableau. Clés d'objet triées par octets
 * (ksort SORT_STRING) - équivalent au tri UTF-16 de C# tant que les clés sont ASCII
 * (elles le sont dans tout le contrat).
 */
final class Jcs
{
    public static function canonical(mixed $node): string
    {
        $out = '';
        self::write($node, $out);
        return $out;
    }

    private static function write(mixed $node, string &$out): void
    {
        if (is_null($node)) { $out .= 'null'; return; }
        if (is_bool($node)) { $out .= $node ? 'true' : 'false'; return; }
        if (is_int($node)) { $out .= (string) $node; return; }
        if (is_float($node)) {
            if (is_finite($node) && floor($node) === $node) { $out .= (string) (int) $node; return; }
            throw new \RuntimeException('Nombre non entier dans le passeport canonique');
        }
        if (is_string($node)) { self::writeString($node, $out); return; }
        if (is_array($node)) { // tableau JSON (liste)
            $out .= '[';
            $first = true;
            foreach ($node as $item) { if (!$first) $out .= ','; $first = false; self::write($item, $out); }
            $out .= ']';
            return;
        }
        if ($node instanceof \stdClass) {
            $props = get_object_vars($node);
            ksort($props, SORT_STRING); // tri par octets = UTF-16 pour clés ASCII
            $out .= '{';
            $first = true;
            foreach ($props as $k => $v) {
                if (!$first) $out .= ',';
                $first = false;
                self::writeString((string) $k, $out);
                $out .= ':';
                self::write($v, $out);
            }
            $out .= '}';
            return;
        }
        throw new \RuntimeException('Type non supporté dans la canonicalisation');
    }

    private static function writeString(string $s, string &$out): void
    {
        $out .= '"';
        $len = strlen($s);
        for ($i = 0; $i < $len; $i++) {
            $c = $s[$i];
            switch ($c) {
                case '"':  $out .= '\\"'; break;
                case '\\': $out .= '\\\\'; break;
                case "\x08": $out .= '\\b'; break;
                case "\x0C": $out .= '\\f'; break;
                case "\n": $out .= '\\n'; break;
                case "\r": $out .= '\\r'; break;
                case "\t": $out .= '\\t'; break;
                default:
                    $o = ord($c);
                    if ($o < 0x20) $out .= sprintf('\\u%04x', $o);
                    else $out .= $c; // ASCII / octets UTF-8 laissés littéraux
            }
        }
        $out .= '"';
    }
}
